pkg/controller/deployment/util/deployment_util.go

```diff
@@ -699,6 +806,64 @@ func IsRollingUpdate(deployment *extensions.Deployment) bool {
 	return deployment.Spec.Strategy.Type == extensions.RollingUpdateDeploymentStrategyType
 }
 
+// IsDeploymentComplete considers a deployment to be complete once its desired replicas equals its
+// updatedReplicas and it doesn't violate minimum availability.
+func IsDeploymentComplete(deployment *extensions.Deployment, newStatus extensions.DeploymentStatus) bool {
```

Comment:
Use pointers for the majority of these - mixing value types and pointers makes it more dangerous that someone thinks they're safe, whereas with pointers they need to be slightly cautious.